src/libasr/pass/pass_list_concat.cpp

```diff
@@ -0,0 +1,220 @@
+
+            // Copy `left_list` contents
+            do_loop_head.m_start = ASRUtils::EXPR(make_IntegerConstant_t(
+                al, loc, 0, int_type));
+            do_loop_head.m_end = ASRUtils::EXPR(ASR::make_IntegerBinOp_t(
+                al, loc, ASRUtils::EXPR(ASR::make_ListLen_t(
+                    al, loc, arg_exprs[0], int_type, nullptr)
+                ), ASR::binopType::Sub,
+                ASRUtils::EXPR(ASR::make_IntegerConstant_t(
+                    al, loc, 1, int_type)
+                ), int_type, nullptr));
+            do_loop_head.m_increment = ASRUtils::EXPR(make_IntegerConstant_t(
+                al, loc, 1, int_type));
+            do_loop_head.loc = loc;
+            Vec<ASR::stmt_t*> loop_body;
+            loop_body.reserve(al, 1);
+            ASR::stmt_t* list_append = ASRUtils::STMT(ASR::make_ListAppend_t(
+                al, loc, arg_expr, ASRUtils::EXPR(ASR::make_ListItem_t(
+                    al, loc, arg_exprs[0], do_loop_head.m_v, item_type, nullptr))));
+            loop_body.push_back(al, list_append);
+            list_concat_stmt = ASRUtils::STMT(ASR::make_DoLoop_t(
+                al, loc, do_loop_head, loop_body.p, loop_body.n));
+            body.push_back(al, list_concat_stmt);
+
+            // Copy `right_list` contents
+            do_loop_head.m_end = ASRUtils::EXPR(ASR::make_IntegerBinOp_t(
+                al, loc, ASRUtils::EXPR(ASR::make_ListLen_t(
+                    al, loc, arg_exprs[1], int_type, nullptr)
+                ), ASR::binopType::Sub,
+                ASRUtils::EXPR(ASR::make_IntegerConstant_t(
+                    al, loc, 1, int_type)
+                ), int_type, nullptr));
+            loop_body.p = nullptr, loop_body.n = 0;
+            loop_body.reserve(al, 1);
+            list_append = ASRUtils::STMT(ASR::make_ListAppend_t(
+                al, loc, arg_expr, ASRUtils::EXPR(ASR::make_ListItem_t(
+                    al, loc, arg_exprs[1], do_loop_head.m_v, item_type, nullptr))));
+            loop_body.push_back(al, list_append);
+            list_concat_stmt = ASRUtils::STMT(ASR::make_DoLoop_t(
+                al, loc, do_loop_head, loop_body.p, loop_body.n));
+            body.push_back(al, list_concat_stmt);
+
+            ASR::asr_t *fn = ASR::make_Function_t(
+            al, loc,
+            /* a_symtab */ list_concat_symtab,
+            /* a_name */ s2c(al, fn_name),
+            /* a_args */ arg_exprs.p,
+            /* n_args */ arg_exprs.n,
+            /* a_body */ body.p,
+            /* n_body */ body.n,
+            /* a_return_var */ arg_expr,
+            ASR::abiType::Source,
+            ASR::accessType::Public, ASR::deftypeType::Implementation,
+            nullptr,
+            false, false, false, false, false,
+            nullptr, 0,
+            nullptr, 0,
+            false);
+            global_scope->add_symbol(fn_name, down_cast<ASR::symbol_t>(fn));
+            return ASR::down_cast<ASR::symbol_t>(fn);
```

Comment:
After applying the suggestions. I get `LCompilersException: visit_DoLoop() not implemented`. So, on the global scope, `DoLoop` is not implemented. So, in this case, we have to create a function scope, right?

Reply:
I think the `do_loop` pass is not getting applied to your function. But it gets applied to vector copy intrinsic optimisation fallback. So probably there is something incorrect here. Can you try to figure it out?

Reply:
Got it, I will try to debug it.